test: add check that too large txs aren't put into orphanage

theStack opened the pull request:
This PR adds test coverage for the following check in `TxOrphanage::AddTx`, where large orphan txs are ignored in order to avoid memory exhaustion attacks:
https://github.com/bitcoin/bitcoin/blob/5abb9b1af49be9024f95fa2f777285c531785d85/src/txorphanage.cpp#L22-L34
Note that this code-path isn't reachable under normal circumstances, as txs larger than `MAX_STANDARD_TX_WEIGHT` are already rejected earlier in the course of doing the mempool standardness checks (see `MemPoolAccept::PreChecks` -> `IsStandardTx` -> `reason = "tx-size";`), so this is only relevant if tx standardness rules are disabled via `-acceptnonstdtxns=1`. The ignore path is checked ~~by asserting the debug log, which is ugly, but as far as I know there is currently no way to access the orphanage entries from the outside~~ via unit test that checks the return value of `AddTx`. As an alternative to adding test coverage, one might consider removing this check altogether (or replacing it with an `Assume`), as it's redundant as explained above.

tdb3 reviewed (approved):
ACK 30aab71063d039d4386f580a031cb5e5e742c363

Nice test addition.  Worked well.

> The ignore path is checked by asserting the debug log, which is ugly, but as far as I know there is currently no way to access the orphanage entries from the outside. As an alternative to adding test coverage, one might consider removing this check altogether (or replacing it with an Assume), as it's redundant as explained above.

RPC already has a way to get mempool transactions (e.g. `getrawmempool` and `getmempoolentry`), but as far as I can tell, not orphans. Unless I'm overlooking something (or there's a history I'm not aware of), might be nice to have a `getorphantxs` (returning orphan txids, etc.) and `getorphan "txid"` (details for an orphan). Orphans do expire relatively quickly, so results from these RPC would be pretty ephemeral, but it might still be worth knowing. I'm taking a look at implementation options.

An added bonus would be that `p2p_orphan_handling` could call `getorphantxs` directly (rather than search the debug log). If others agree that these new RPCs are worthwhile, then the RPCs should be implemented in a separate/followup PR, and in my opinion shouldn't impact this PR getting merged (an adjustment of the test could be done later).

tdb3 commented:
> RPC already has a way to get mempool transactions (e.g. `getrawmempool` and `getmempoolentry`), but as far as I can tell, not orphans. Unless I'm overlooking something (or there's a history I'm not aware of), might be nice to have a `getorphantxs` (returning orphan txids, etc.) and `getorphan "txid"` (details for an orphan). Orphans do expire relatively quickly, so results from these RPC would be pretty ephemeral, but it might still be worth knowing. I'm taking a look at implementation options.

Draft PR opened https://github.com/bitcoin/bitcoin/pull/30793

danielabrozzoni reviewed (approved):
tACK 30aab71063d039d4386f580a031cb5e5e742c363

glozow reviewed (commented):
Concept ACK, but this should be a unit test, not a functional test looking for logs. We could accidentally delete the `return false;` line but keep the `LogPrint` line, and this test would pass.

> so this is only relevant if tx standardness rules are disabled via -acceptnonstdtxns=1. 

Similarly, it's a bit icky to have to disable standardness checks to test this.

A unit test in orphanage_tests.cpp could directly create a large tx and call `AddTx`. With #30110, we can also test by calling `MempoolRejectedTx` with a large transaction and `TX_MISSING_INPUTS`.

The harder part there is adding a utility function to bulk transactions in our unit testing framework, but I think that would be quite helpful in the long run, so I'd gladly review that.

theStack committed:
test: add `BulkTransaction` helper to unit test transaction utils

The padding method used matches the one used in MiniWallet,
`MiniWallet._bulk_tx`.

theStack committed:
test: add check that large txs aren't put into orphanage

glozow reviewed (commented):
ACK 66d13c870284327abc89d36c0b5cc5f58e96f570

tdb3 reviewed (approved):
re-ACK 66d13c870284327abc89d36c0b5cc5f58e96f570

Nice unit test addition.
